app/javascript/packs/spinner-button.js

```diff
@@ -0,0 +1,31 @@
+/**
+ * @typedef SpinnerButtonElements
+ *
+ * @prop {HTMLDivElement} wrapper
+ * @prop {HTMLImageElement} spinner
+ * @prop {HTMLButtonElement|HTMLInputElement|HTMLLinkElement} button
+ */
+
+export class SpinnerButton {
+  constructor(wrapper) {
+    /** @type {SpinnerButtonElements} */
+    this.elements = {
+      wrapper,
+      spinner: wrapper.querySelector('.spinner-button__spinner'),
+      button: wrapper.querySelector('a,button:not([type]),[type="submit"]'),
+    };
+
+    this.bindEvents();
+  }
+
+  bindEvents() {
+    this.elements.button.addEventListener('click', () => this.showSpinner());
+  }
+
+  showSpinner() {
+    this.elements.wrapper.classList.add('spinner-button--spinner-active');
+    this.elements.spinner.classList.remove('usa-sr-only');
+  }
+}
+
+[...document.querySelectorAll('.spinner-button')].forEach((wrapper) => new SpinnerButton(wrapper));
```

Comment:
> this line makes this a side-effecty import

I agree it does. I've come to be okay to consider "packs" as an integrating point to be side-effecty. As an improvement, I think an alternative I'd like is to move the class definition out, to either `app/javascript/app` or `app/javascript/packages`. An inline script could help avoid there being this intermediary, and also move the initialization closer to the markup, but:

1. To the original bug report, ES2015 / import syntax is not transpiled and would error.
2. To my point below about your second question, an inline script would ideally only concern the _current_ partial, not activate all spinner buttons. The mere act of finding that element would be a bit troublesome, either requiring a consumer to pass in an identifier, generating a unique identifier automatically, or using something like `document.currentScript.previousElementSibling`.

>Also if we ever included the spinner ERB partial twice, the second spinner wouldn't be activated right?

No, they'd both be activated. Because it uses `javascript_pack_tag_once` (vs. `javascript_pack_tag`, see #4221), the script isn't actually printed here, it's [printed at the end of the page](https://github.com/18F/identity-idp/blob/60a220cd4b6ad305dc7265c21d833c59f7c7d41d/app/views/layouts/base.html.erb#L90). So this code is run once, and should capture any and all instances of the "component" in the page.

---

An adaptation of your suggestion could be something similar to...

```erb
<% id = SecureRandom.uuid %>
<%= tag.div id: id, class: 'spinner-button' %>
<%= javascript_pack_tag 'spinner-button' %>
<%= nonced_javascript_tag do %>
  new window.LoginGov.SpinnerButton(document.getElementById(<%= id.to_json %>));
<% end %>
```

It wouldn't work well in this componentized, load-once, initialize-once model though. If we wanted something like that, we could create something to buffer inline scripts to be spat out alongside the `render_javascript_pack_once_tags`. Using a bit of my own history as an analogy, I view `javascript_pack_tag_once` as something equivalent to WordPress' [`wp_enqueue_script`](https://developer.wordpress.org/reference/functions/wp_enqueue_script/), where this would be something akin to a [`wp_add_inline_script`](https://developer.wordpress.org/reference/functions/wp_add_inline_script/).

Reply:
Oh right I forgot the nonced tags are added at the end --- this seems much simpler as-is then. Thanks!